cyclic_boosting/generic_loss.py

```diff
@@ -0,0 +1,678 @@
+from __future__ import absolute_import, division, print_function
+
+import abc
+import logging
+import warnings
+
+import numpy as np
+import six
+import sklearn.base
+from scipy.optimize import minimize
+from scipy.stats import beta
+
+from cyclic_boosting.base import CyclicBoostingBase, gaussian_matching_by_quantiles
+from cyclic_boosting.link import LogLinkMixin, IdentityLinkMixin, LogitLinkMixin
+from cyclic_boosting.utils import continuous_quantile_from_discrete, get_X_column
+
+_logger = logging.getLogger(__name__)
+
+
+@six.add_metaclass(abc.ABCMeta)
+class CBGenericLoss(CyclicBoostingBase):
+    """
+    A generic loss, to be defined in the respective subclass, is minimized in
+    each bin of each feature. While binning, feature cycles, smoothing, and
+    iterations work in the same way as usual in Cyclic Boosting, the
+    minimization itself is performed via ``scipy.optimize.minimize`` (instead
+    of an analytical solution like, e.g., in ``CBPoissonRegressor``,
+    ``CBNBinomRegressor``, or ``CBLocationRegressor``).
+    """
+
+    def precalc_parameters(self, feature, y, pred):
```

Comment:
Would be nice to use typing throughout here as well 😈 Otherwise I can do this later too

Reply:
Haha, fair enough. Let me quickly do that :).